siemens_cp1616/src/siemens_cp1616_io_controller.cpp

```diff
@@ -0,0 +1,522 @@
+    return (int)error_code;
+  }
+
+  if(cp_handle_ == valid_cp_handle)  //check if cp_handle_ still valid
+  {
+    //wait for a callback_for_mode_change_indication
+    while(!sem_mod_change_){
+    usleep(WAIT_FOR_CALLBACKS_PERIOD);
+    }
+
+    setSemModChange(0);
+  }
+
+  //check if the current mode is correct
+  if(cp_current_mode_ != requested_mode)
+  {
+    ROS_ERROR("Not able to set required mode: ERROR another mode recieved");
+    PNIO_close(cp_handle_);
+  }
+  else
+    ROS_INFO("Changing IO_controller mode: done");
+
+  return (int)error_code;
+}
+
+int Cp1616IOController::updateCyclicInputData()
+{
+  PNIO_UINT32 error_code;
+  PNIO_UINT32 bytes_read;
+
+  unsigned int i;
+
+  for(i = 0; i < input_module_count_; i++)
+  {
+    error_code = PNIO_data_read(
+                    cp_handle_,                             // handle
+                    &input_data_address_[i],                // pointer to device input address
+                    input_data_length_[i],                  // length in bytes of input
+                    &bytes_read,                            // number of bytes read
+                    &input_data_[i][0],                     // pointer to input data row
+                    cp_local_state_,                        // local status
+                    (PNIO_IOXS*)&(input_data_state_[i]));   // remote status
+
+    if(error_code != PNIO_OK)
+      ROS_DEBUG("PNIO_read_data (PNIO_CBE_DEV_ACT_CONF,..) returned 0x%x", (int)error_code);
+    
+    /* Print input_data to terminal (debugging purposes)  
+    std::cout << "Input  m." << i << std::dec
+               << " [I: " << input_data_address_[i].u.Addr
+               << " - "  << input_data_address_[i].u.Addr + input_data_length_[i]-1
+               << "]: ";
+
+    for(int i = 0; i < input_data_length_[i]; i++)
+        std::cout <<  std::setfill(' ') << std::setw(2) << std::hex << (int)input_data_[i][i] << " ";
+
+    std::cout << std::endl;
+    */
+  }
+
+  return (int)error_code;
+}
+
+int Cp1616IOController::updateCyclicOutputData()
+{
+  PNIO_UINT32 error_code;
+  unsigned int i;
+
+  for(i = 0; i < output_module_count_; i++)
+  {
+    error_code = PNIO_data_write(
+                    cp_handle_,                             // handle
+                    &output_data_address_[i],               // pointer to device output address
+                    output_data_length_[i],                 // length in bytes of output
+                    &output_data_[i][0],                    // pointer to output data row
+                    cp_local_state_,                        // local status
+                    (PNIO_IOXS*)&(output_data_state_[i]));  // remote status
+
+    if(error_code != PNIO_OK)
+        ROS_DEBUG("PNIO_read_data (PNIO_CBE_DEV_ACT_CONF,..) returned 0x%x", (int)error_code);
+    
+    /* Print output_data_ to terminal (debugging purposes)
+    std::cout << "Output m." << i << std::dec
```

Comment:
This prints aligned data to the terminal in the loop. I was thinking about using ROS_DEBUG but I was not able to reach the same appearance because of [INFO] [TIME STAMP] part of ROS_DEBUG message. It is removed for now but for debugging purposes it would be useful to keep it there.